File: LC/python/Sort List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param head, a ListNode
    # @return a ListNode
    #merge sort O(nlogn) Time
    def sortList(self, head):
        if head == None or head.next == None:
            return head
        slow = head
        quick = head
        
        while quick.next and quick.next.next:
            slow = slow.next
            quick = quick.next.next
        head1 = head
        head2 = slow.next
        slow.next = None
        
        head1 = self.sortList(head1)
        head2 = self.sortList(head2)
        head = self.merge(head1,head2)
        return head
    
    def merge(self,head1,head2):
        if head1 == None:
            return head2
        if head2 == None:
            return head1
        
        dummy = ListNode(0)
        p = dummy
        
        while head1 and head2:
            if head1.val<head2.val:
                p.next = head1
                head1 = head1.next
                p = p.next
            else:
                p.next = head2
                head2 = head2.next
                p = p.next
            
        if head1 == None:
            p.next = head2
        if head2 == None:
            p.next = head1
        return dummy.next
        
        

File: LC/java/Sort List.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode sortList(ListNode head) {
        if(head==null || head.next==null){
            return head;
        }
        
        ListNode slow,fast,head1,head2;
        slow = fast = head1 = head;
        while(fast.next!=null && fast.next.next!=null){
            slow = slow.next;
            fast = fast.next.next;
        }
        
        head2 = slow.next;
        slow.next = null;
        
        head1 = sortList(head1);
        head2 = sortList(head2);
        head = merge(head1,head2);
        
        return head;
    }
    
    public ListNode merge(ListNode head1, ListNode head2){
        ListNode tmp = new ListNode(0);
        ListNode res = tmp;
        while(head1!=null && head2!=null){
            if(head1.val < head2.val){
                res.next = head1;
                head1 = head1.next;
                res = res.next;
            }
            else{
                res.next = head2;
                head2 = head2.next;
                res = res.next;
            }
        }
        
        if(head1==null){
            res.next = head2;
        }
        else{
            res.next = head1;
        }
        return tmp.next;
    }
   
}